File: eplant_poplar/cgi-bin/get_interactions.py
#!/usr/bin/python3
"""
This program gets interactions between interactions including PDIs
Author: Asher
Date: May, 2019
Example: http://www.bar.utoronto.ca/eplant_poplar/cgi-bin/get_interactions_dapseq.py?locus=Potri.001G001300
"""
import MySQLdb
import cgi
import re
import json
from sys import exit


def connect():
    """
    Connect to the database
    :return: Database connection
    """

    try:
        conn = MySQLdb.connect(host='localhost', user='InteractionsUser', passwd='InteractionsUser', db='poplar_interactions')
    except MySQLdb.Error:
        output_error('Can not connect to database!')

    return conn


def first_query(conn, agi):
    """
    Get the results of the first query (Proteins only).
    :param conn: MySQL connection
    :param str agi: AGI ID
    :return: An Hash of protein
    """

    proteins = {}
    results = []

    # Prepare query
    cursor = conn.cursor()
    sql = 'SELECT Protein1, Protein2 FROM interactions WHERE aiv_index < 2 AND Protein1 = %s'

    # Run query
    try:
        cursor.execute(sql, (agi, ))
        results = cursor.fetchall()
    except MySQLdb.Error as e:
        conn.close()
        output_error('Could not query database.')

    # Check if there are no results
    if len(results) == 0:
        print({})
        conn.close()
        exit()

    # Get data from results
    for row in results:
        # Add query proteins
        if row[0] not in proteins:
            proteins[row[0]] = ''

        # Add interactions
        if row[1] not in proteins:
            proteins[row[1]] = ''

    return proteins


def second_query(conn, agis):
    """
    This function runs the second query of proteins
    :param conn: connection
    :param agis: hash of agis
    :return: list of dictionaries of proteins and dna
    """
    proteins = {}
    results = []

    # Query PPIs
    cursor = conn.cursor()
    sql = 'SELECT Protein1, Protein2, aiv_index, cv FROM interactions WHERE aiv_index < 2 AND Protein1 IN ("' + '", "'.join(
        agis.keys()) + '")'

    # Run query
    try:
        cursor.execute(sql)
        results = cursor.fetchall()
    except MySQLdb.Error:
        conn.close()
        output_error('Could not query database.')

    # Check if there are no results
    if len(results) == 0:
        conn.close()
        print({})
        exit()

    for row in results:
        # Protein/DNA check
        interaction = row[0] + "\t" + row[1]
        if interaction not in proteins:
            proteins[interaction] = row

    return proteins


def hash_filter(first_hits, second_hits):
    """
    This function filters the second network using dictionaries
    :param first_hits: Network from first query
    :param second_hits: Network from second query
    :return: filtered network
    """

    protein_filtered = {}

    # Now filter

    # Protein
    for interaction, row in second_hits.items():
        p1, p2 = interaction.split("\t")
        if p2 in first_hits:
            protein_filtered[interaction] = row

    return protein_filtered


def remove_duplicates(data):
    """
    This function removes duplicates from the interactome
    :param data: Hash filtered data
    :return: data with duplicates removed
    """

    uniq_proteins = dict()
    final_data = []

    # Proteins
    for interaction, row in data.items():
        p1, p2 = interaction.split("\t")
        query = p2 + "\t" + p1
        if query in uniq_proteins:
            pass
        else:
            uniq_proteins[interaction] = row

    # Final data is array of array like the one returned by MySQL
    for interaction, row in uniq_proteins.items():
        final_data.append(row)

    return final_data


def add_PDIs(conn, agi, data):
    """
    Add PDIs in the result as a final step
    :param conn: MySQL connection
    :param agi: Locus gene
    :param data: Dataset so far
    :return: Full data set.
    """
    results = []

    # Query PPIs
    cursor = conn.cursor()
    sql = 'SELECT Protein1, Protein2, aiv_index, cv FROM interactions WHERE aiv_index = 2 AND Protein1 = %s'

    # Run query
    try:
        cursor.execute(sql, (agi,))
        results = cursor.fetchall()
    except MySQLdb.Error:
        conn.close()
        output_error('Could not query database.')

    # Check if there are no results
    if len(results) == 0:
        return data

    for row in results:
        data.append(row)

    return data


def output_data(recursive, agi, data):
    """
    This functions outputs interactions.
    :param recursive: If there is interaction between interactors
    :param agi: Gene ID of query
    :param data: rows returned from MySQL
    :return:
    """

    final_output = dict()
    final_output[agi] = []

    for row in data:
        output = dict()
        output['source'] = row[0]
        output['target'] = row[1]
        output['index'] = str(row[2])

        if row[3] is None:
            output['interolog_confidence'] = None
        else:
            output['interolog_confidence'] = float(row[3])

        # Stuff not in Poplar yet
        #if row[4] is None:
        #    output['interolog_confidence'] = None
        #else:
        #    output['correlation_coefficient'] = None

        ## Set default for published
        #if row[5] is None:
        #    output['published'] = False
        #    output['reference'] = "None"
        #else:
        #    output['published'] = True
        #    output['reference'] = row[5]

        ## Now If it a predicted DNA, then set published = false
        #if row[3] is not None:
        #    if row[2] == 2 and (0 < row[3] < 1):
        #        output['published'] = False

        output['published'] = True
        output['correlation_coefficient'] = None
        output['reference'] = "Biogrid";
        final_output[agi].append(output)

    if recursive:
        final_output[agi].append({'recursive': 'true'})
    else:
        final_output[agi].append({'recursive': 'false'})

    print(json.dumps(final_output))
    exit()


def is_locus_valid(locus):
    """
    This function checks if locus is valid
    :param locus: string Input AGI
    :return: boolean. True if worked. False if format is not right.
    """
    if re.match(r"^Potri\.\d*G\d+$", locus, re.IGNORECASE):
        return True
    else:
        return False


def output_error(message):
    """
    This function outputs error messages
    :param message: string
    :return:
    """
    output = dict()
    output['status'] = 'fail'
    output['error'] = message
    print(json.dumps(output))
    exit()


def main():
    """
    The main program.

    :return:
    """

    print('Content-type: application/json\n')

    try:
        arguments = cgi.FieldStorage()
        locus = arguments['locus'].value
    except Exception:
        output_error('Could not read locus parameter of GET request.')

    # Validate locus
    if is_locus_valid(locus):
        pass
    else:
        output_error('Locus is invalid.')

    # Connect to database
    conn = connect()

    # Run the first query
    first_hits = first_query(conn, locus)

    # Run the second query using protein in first queries result
    second_hits = second_query(conn, first_hits)

    # Hash filter
    hash_filtered_second_hits = hash_filter(first_hits, second_hits)
    final_data = remove_duplicates(hash_filtered_second_hits)

    # Add PIDs
    final_output = add_PDIs(conn, locus, final_data)
    conn.close()

    # Output the final data, So far, recursive is set to true.
    # Are we not doing interactions between interactors?
    output_data(True, locus, final_output)


if __name__ == '__main__':
    main()

File: eplant_sunflower/cgi-bin/get_interactions.php
<?php
/**
* Interactions webservice.
*
* Retrieve Interactions and SUBA3 data from the BAR databases for the use in ePlant 2.0 only.
* 
* @category	WebServices
* @Author	Asher
* @MySQL	Matt Ierullo
* @Date		July, 2014
* @Usage	http://bar.utoronto.ca/webservices/new_get_interactions.php?request=[{"agi":"At1g02130"},{"agi":%"At4g23810"},{"agi":"At3g01090"}]	
*/

// The Interactions Class
class Interactions
{
	// Constructor
	function __construct($JSONData) {
		$this->JSONData = $JSONData;
	}
	
	// Is the JSON valid?
	public function validateData() {
		if (json_decode($this->JSONData)) {
			$this->JSONData = json_decode($this->JSONData, true);
			return 1;
		} else {
			return 0;
		}
	}

	// Get the interactions
	public function getInteractions() {
		// Connect to MySQL
		$mysqli = $this->connect();

		// Now get interactions data
		$interactions = $this->execute($mysqli);

		// Disconnect from MySQL
		$this->disconnect($mysqli);
		
		return $interactions;
	}

	// Get interactions data
	private function execute($mysqli) {
		$finalData = array();
		$getRecursive = False;	// Query recursively is true by defalt
		
		for ($i = 0; $i < count($this->JSONData); $i++) {
			if (isset($this->JSONData[$i]["agi"])) {
				$gene = $this->JSONData[$i]["agi"];
				// Security checks
				$gene = $this->cleanInput($gene);
				
				// For MySQL
				$gene = $mysqli->real_escape_string($gene);

				// Run the first query
				if ($interactors = $mysqli->query("SELECT Protein1, Protein2, Quality, Pcc, Bind_id, aiv_index FROM interactions WHERE Protein1 = '$gene'")) {
					// Check if there are too many interactions. If yes, don't recursive query
					if ($interactors->num_rows < 50) {
						$getRecursive = True;
						// Get data
						$listInteractors = array($gene);
						while ($row = $interactors->fetch_assoc()) {
							array_push($listInteractors, $row["Protein2"]);
						}
					} else {
						$getRecursive = False;
						// Get and Store data
						$data = array();
						while ($row = $interactors->fetch_assoc()) {

							// Assign publised and reference
							if (is_null($row["Bind_id"])) {
								$published = false;
								$reference = 'None';
							} else {
								if ($row["aiv_index"] == "2" && $row["Bind_id"] != "PubMed22037706") {
									$published = false;
								} else {
									$published = true;
								}
								$reference = $row["Bind_id"];
							}

							// Fix predicted PDI reference
							if ($row["aiv_index"] == "2" && $row["Bind_id"] == "PubMed25352272") {
								$published = true;
							}

							// Push everything to data array
							array_push($data, array(
								"source" => $row["Protein1"],
								"target" => $row["Protein2"],
								"interolog_confidence" => $row["Quality"],
								"correlation_coefficient" => $row["Pcc"],
								"published" => $published,
								"reference" => $reference,
								"index" => $row["aiv_index"])
							);
						}
						array_push($data, array("recursive" => "false"));
					}
					$interactors->close();
				}

				if ($getRecursive) {
					// Now Build the second query
					$query = "SELECT Protein1, Protein2, Quality, Pcc, Bind_id, aiv_index FROM interactions WHERE (Protein1='" . $gene . "' AND Protein2='" . $gene . "')";
					foreach($listInteractors as $p1) {
						foreach($listInteractors as $p2) {
							if (! preg_match("/(Protein1='".$p2."' AND Protein2='".$p1."')/", $query) or preg_match("/(Protein1='".$p1."' AND Protein2='".$p2."')/", $query)) {
								$query = $query . " OR (Protein1='" . $p1 . "' AND Protein2='" . $p2 . "')";
							}
						}
					}

					// Now Run the query
					$data = array();
					if ($interactors = $mysqli->query($query)) {
											
						// Get Data
						while ($row = $interactors->fetch_assoc()) {
							// Assign publised and reference
							if (is_null($row["Bind_id"])) {
								$published = false;
								$reference = 'None';
							} else {
								$published = true;
								$reference = $row["Bind_id"];
							}

							// Push to data array
							array_push($data, array(
								"source" => $row["Protein1"],
								"target" => $row["Protein2"],
								"interolog_confidence" => $row["Quality"],
								"correlation_coefficient" => $row["Pcc"],
								"published" => $published,
								"reference" => $reference,
								"index" => $row["aiv_index"])
							);
						}
						array_push($data, array("recursive" => "true"));
						$interactors->close();
					}
				} 

				if (isset($data)) {	
					$finalData[$gene] = $data;
				} else {
					$this->end();
				}
			} else {
				$this->end();
			}
		}
		return $finalData;
	}

	// Clean input
	private function cleanInput($data) {
		$data = trim($data);
		$data = stripslashes($data);
		$data = htmlspecialchars($data);
		if (preg_match("/^At.g\d{5}$/i", $data)) {
			return $data;	
		} else {
			$this->end();
		}
	}

	// Connect to database
	private function connect() {
		$mysqli = new mysqli("localhost", "InteractionsUser", "InteractionsUser", "interactions");
		
		if ($mysqli->connect_errno) {
			$this->end();
		}
		return $mysqli;
	}

	// Disconnect from database
	private function disconnect($mysqli) {
		$mysqli->close();
	}
	
	// Exit with error
	private function end() {
		header("Content-type: application/json");
		$response["Error"] = "An error has occured";
		echo json_encode($response);
		exit;
	}
}

// The main program.
if (isset($_GET['request'])) {
	// Create a new object
	$interactionsObj = new Interactions($_GET['request']);

	header("Content-type: application/json");
	if ($interactionsObj->validateData()) {
		$response = $interactionsObj->getInteractions();
		echo json_encode($response);
	} else {
		$response["Error"] = "Invalid input data";
		echo json_encode($response);
	}
}

